### PR TITLE
Implement action sheet for outbox messages

### PR DESCRIPTION
**Pull request by kunall17**
Long press menu for deleting outbox message or retrying to send message

### CODE REVIEW

**Comment by smarx** [comment]
*Automated message from Dropbox CLA bot*

@kunall17, it looks like you've already signed the Dropbox CLA. Thanks!

**Comment by borisyankov** [comment]
The points @Sam1301 made on the previous (already merged) PR are quite valid, and now with this PR I am surer of that...
The more 'container' components we have the more duplication we will have...
We already have a 'Delete' item in the other menu, it just needs enhancing to catch if a message is in the outbox.

And we will not have a 'Retry' menu item. It just have to retry automatically.

**Comment by kunall17** [comment]
Ok i'll merge them both, but it might be messy and filled with conditional codes! 

> And we will not have a 'Retry' menu item. It just have to retry automatically.

I think we should have this for now (atleast) as the auto sending might not be perfect and needs some more feedback/improvements 
Thoughts @Sam1301 @vishwesh3 ?

**Comment by borisyankov** [comment]
Then, auto-sending of failed messages will be the highest priority task. The idea of an outbox is 80% about preserving what a user sent but was not received by the backend, and 20% about other stuff like the preview. The main task is not done until it can do that.

**Comment by Sam1301** [comment]
I haven't looked closely into the outbox functionality/implementation so can't comment on this :).

**Comment by kunall17** [comment]
Fixes #1032 , @borisyankov LGTM sooner!

File: src/utils/message.js [hunk 1]
```diff
@@ -37,6 +37,10 @@ export const isSameRecipient = (message1: Message, message2: Message): boolean =
         message1.display_recipient.toLowerCase() === message2.display_recipient.toLowerCase() &&
         message1.subject.toLowerCase() === message2.subject.toLowerCase()
       );
+    case 'outbox': {
+      // If both are an outbox message return true
```
Comment (borisyankov) [inline src/utils/message.js]: Code is pretty clear without the comment.

File: src/message/MessageContainer.js [hunk 2]
```diff
@@ -87,21 +88,38 @@ class MessageContainer extends PureComponent {
     this.showActionSheet({ options, cancelButtonIndex: options.length - 1, callback });
   };
 
+  handleOutboxMessageLongPress = () => {
+    const { actions, message } = this.props;
+    const callback = buttonIndex => {
+      switch (buttonIndex) {
```
Comment (borisyankov) [inline src/message/MessageContainer.js]: This is not the place we handle the action sheet actions.

File: src/message/MessageContainer.js [hunk 3]
```diff
@@ -87,21 +88,38 @@ class MessageContainer extends PureComponent {
...
+        case 0: {
+          actions.deleteOutboxMessage(message.timestamp);
+          break;
+        }
+        default:
+          break;
+      }
+    };
+    const options = ['Delete message', 'Cancel'];
+    this.showActionSheet({ options, cancelButtonIndex: options.length - 1, callback });
```
Comment (borisyankov) [inline src/message/MessageContainer.js]: Most of the actions are relevant for an outbox message, why have a specific action sheet for outbox?

File: src/message/MessageContainer.js [hunk 4]
```diff
@@ -87,21 +88,38 @@ class MessageContainer extends PureComponent {
...
     const MessageComponent = isBrief ? MessageBrief : MessageFull;
 
     return (
       <MessageComponent
         message={message}
         twentyFourHourTime={twentyFourHourTime}
         ownEmail={auth.email}
-        onLongPress={this.handleLongPress}
-        starred={this.isStarred(message)}
+        onLongPress={isOutbox ? this.handleOutboxMessageLongPress : this.handleLongPress}
```
Comment (borisyankov) [inline src/message/MessageContainer.js]: Not here.

File: src/message/MessageContainer.js [hunk 5]
```diff
@@ -87,21 +88,38 @@ class MessageContainer extends PureComponent {
...
+        starred={isOutbox ? undefined : this.isStarred(message)}
         realm={auth.realm}
         auth={auth}
         actions={actions}
         handleLinkPress={this.handleLinkPress}
+        isOutbox={isOutbox}
       />
     );
   }
```
Comment (borisyankov) [inline src/message/MessageContainer.js]: You need to think more in the direction of 'separation of concerns'.
Adding more and more logic to a single file/component usually breaks it.
MessageContainer ideally should have no idea about the outbox concept.
In fact, outbox should be as similar to non-outbox as possible, so very little 'ifs' are needed.

File: src/message/MessageFull.js [hunk 6]
```diff
@@ -39,8 +39,9 @@ class MessageFull extends PureComponent {
     auth?: Auth,
     message: Object,
     onLongPress?: () => void,
-    isNotYetSent?: boolean,
     handleLinkPress?: string => void,
+    onLongPress: () => void,
+    isOutbox?: boolean,
```
Comment (borisyankov) [inline src/message/MessageFull.js]: Can we determine 'isOutbox' later?

File: src/outbox/outboxReducers.js [hunk 7]
```diff
@@ -9,6 +9,7 @@ const reducer = (state: OutboxState = initialState, action: Action): OutboxState
     case MESSAGE_SEND: {
       return [...state, { ...action.params }];
     }
+    case DELETE_OUTBOX_MESSAGE:
```
Comment (Sam1301) [inline src/outbox/outboxReducers.js]: wondering whats the use case here? Is it intended to change the state like the `EVENT_NEW_MESSAGE` case?
Also might be useful to add a test for this?

Comment (kunall17) [inline src/outbox/outboxReducers.js]: When a new message comes in we need to check if the id matches then delete from reducer and similar thing is done in the case of `DELETE_OUTBOX_MESSAGE` hence we can have the same block!

**Comment by kunall17** [comment]
@borisyankov Updated!

File: src/message/MessageFull.js [hunk 8]
```diff
@@ -75,20 +74,20 @@ class MessageFull extends PureComponent {
           />
           <View style={styles.contentWrapper}>
             <ScrollView style={styles.inner}>
-              <TouchableWithoutFeedback onLongPress={onLongPress}>
+              <TouchableOpacity onLongPress={onLongPress}>
```
Comment (kunall17) [inline src/message/MessageFull.js]: I think TouchableOpacity looks better when onLongPress, hence I used this and extracted this to another commit, if you don't agree push without this commit!

Comment (borisyankov) [inline src/message/MessageFull.js]: Yeah, I don't think it is good, we had TouchableOpacity before. Most apps I tested had no feedback on long press.

**Comment by kunall17** [comment]
Made the grouping same as normal message, and some enhancements @borisyankov 

 simulator screen shot 17-aug-2017 9 46 53 am

File: src/message/MessageContainer.js [hunk 9]
```diff
@@ -104,7 +104,7 @@ class MessageContainer extends PureComponent {
         twentyFourHourTime={twentyFourHourTime}
         ownEmail={auth.email}
         onLongPress={this.handleLongPress}
-        starred={this.isStarred(message)}
+        starred={message.isOutbox ? undefined : this.isStarred(message)}
```
Comment (borisyankov) [inline src/message/MessageContainer.js]: This should be handled in `isStarred`

Comment (borisyankov) [inline src/message/MessageContainer.js]: But also, likely does not need to be handled at all.

As this will always return false:
```js
isStarred(message: Object) {
    const { flags } = this.props;
    return message.id in flags.starred;
  }
```

File: src/message/MessageTags.js [hunk 10]
```diff
@@ -45,7 +45,7 @@ export default class MessageTags extends PureComponent {
           <View style={styles.tag}>
             <Label style={styles.text} text={'starred'} />
           </View>}
-        {isNotYetSent &&
+        {isOutbox &&
```
Comment (borisyankov) [inline src/message/MessageTags.js]: Show a small spinner instead of this text (really tiny spinner).

File: src/message/messageActionSheet.js [hunk 11]
```diff
@@ -73,6 +73,9 @@ type AuthMessageAndNarrow = {
   narrow: [],
 };
 
+const isAnOutboxMessage = ({ message }: Message): boolean =>
+  message.isOutbox && message.isOutbox === true;
```
Comment (borisyankov) [inline src/message/messageActionSheet.js]: `message.isOutbox === true' will never execute, as ` message.isOutbox` is already true. In that case, though, you totally do not need this extracted as a function.

File: src/message/messageActionSheet.js [hunk 12]
```diff
@@ -88,17 +91,22 @@ const reply = ({ message, actions, auth, currentRoute }: MessageAndDoNarrowType)
...
 const copyToClipboard = async ({ auth, message }: AuthAndMessageType) => {
-  const rawMessage = await getSingleMessage(auth, message.id);
+  const rawMessage = isAnOutboxMessage({ message })
+    ? message.content
+    : await getSingleMessage(auth, message.id);
   Clipboard.setString(rawMessage);
   showToast('Message copied!');
 };
 
+const isSentMessage = ({ message }: Message): boolean => !isAnOutboxMessage({ message });
```
Comment (borisyankov) [inline src/message/messageActionSheet.js]: I guess that is also not needed as it is equal to `!message.isOutbox`

File: src/message/messageActionSheet.js [hunk 13]
```diff
@@ -88,17 +91,22 @@ const reply = ({ message, actions, auth, currentRoute }: MessageAndDoNarrowType)
...
+
 const editMessage = async ({ message, actions }: MessageAuthAndActions) => {
   actions.startEditMessage(message.id);
 };
 
-const deleteMessage = async ({ auth, message }: AuthAndMessageType) => {
-  deleteMessageApi(auth, message.id);
+const deleteMessage = async ({ auth, message, actions }: MessageAuthAndActions) => {
+  if (isAnOutboxMessage({ message })) actions.deleteOutboxMessage(message.timestamp);
```
Comment (borisyankov) [inline src/message/messageActionSheet.js]: Just do it with the `{` and `}` after the `if` and `else`.

File: src/message/renderMessages.js [hunk 14]
```diff
@@ -2,49 +2,34 @@
...
 import { isSameDay } from '../utils/date';
 
 export default (messages: Message[], narrow: Narrow) => {
   const list: Object[] = [];
   let prevItem;
+  const showHeader = !isPrivateOrGroupNarrow(narrow) && !isTopicNarrow(narrow);
 
   for (const item of messages) {
-    const isOutbox = !item.id;
+    const isOutbox = 'isOutbox' in item && item.isOutbox === true;
```
Comment (borisyankov) [inline src/message/renderMessages.js]: You should create normal looking messages in the selector. This function should not be aware of outbox at all.

File: src/message/renderMessages.js [hunk 15]
```diff
@@ -2,49 +2,34 @@
...
-          message={
-            isOutbox
-              ? {
-                  ...item,
-                  display_recipient: extractedOutboxData[1],
-                  subject: extractedOutboxData[2],
-                }
-              : item
-          }
+          key={`header${isOutbox ? item.timestamp : item.id}`}
```
Comment (borisyankov) [inline src/message/renderMessages.js]: Would it work if you just set the item.id to the timestamp in the selector? Then you don't need to have any special handling here.

File: src/message/renderMessages.js [hunk 16]
```diff
@@ -54,23 +39,10 @@ export default (messages: Message[], narrow: Narrow) => {
...
-      );
-      prevItem = item;
-      continue; // eslint-disable-line
-    }
-
     list.push(
       <MessageContainer
-        type="message"
-        key={item.id}
+        type={isOutbox ? 'outbox' : 'message'}
```
Comment (borisyankov) [inline src/message/renderMessages.js]: Same here. The type should be 'message' and outbox-iness should be determined in other way.

File: src/outbox/__tests__/outboxReducer-test.js [hunk 17]
```diff
@@ -36,6 +36,25 @@ describe('outboxReducers', () => {
 
       expect(actualState).toEqual(expectedState);
     });
+
+    test('add a same timestamp message to outbox', () => {
```
Comment (borisyankov) [inline src/outbox/__tests__/outboxReducer-test.js]: Good, you are adding tests 👍 
Likely a good idea to add few to the selectors.

File: src/utils/message.js [hunk 18]
```diff
@@ -38,7 +38,7 @@ export const isSameRecipient = (message1: Message, message2: Message): boolean =
         message1.subject.toLowerCase() === message2.subject.toLowerCase()
       );
     case 'outbox': {
-      return message2.type === 'outbox';
+      return message2.isOutbox && message2.isOutbox === true;
```
Comment (borisyankov) [inline src/utils/message.js]: No need for the second part after the `&&`.

File: src/utils/narrow.js [hunk 19]
```diff
@@ -131,11 +131,26 @@ export const narrowFromMessage = (message: Message, email: string) => {
 
 export const extractTypeToAndSubjectFromNarrow = (
   narrow: Narrow,
-): ['private' | 'stream', string, string] => {
+): { type: 'private' | 'stream', display_recipient: string, subject: string } => {
```
Comment (borisyankov) [inline src/utils/narrow.js]: We use `this_casing` only for data coming from the API directly (we might even revisit that). This should be `displayRecipient` or just `recipient`.

File: src/utils/narrow.js [hunk 20]
```diff
@@ -131,11 +131,26 @@ export const narrowFromMessage = (message: Message, email: string) => {
...
+    return { type: 'private', display_recipient: narrow[0].operand, subject: '' };
   } else if (isStreamNarrow(narrow)) {
-    return ['stream', narrow[0].operand, '(no topic)'];
+    return { type: 'stream', display_recipient: narrow[0].operand, subject: '(no topic)' };
   }
-  return ['stream', narrow[0].operand, narrow[1].operand];
+  return { type: 'stream', display_recipient: narrow[0].operand, subject: narrow[1].operand };
+};
+
+export const compareNarrows = (narrow1: Narrow, narrow2: Narrow): boolean => {
```
Comment (borisyankov) [inline src/utils/narrow.js]: What is the purpose of this function? Looks unused.

File: src/utils/narrow.js (hunk 19 above)
Comment (kunall17) [inline src/utils/narrow.js]: Message object has `display_recipient` so I used it accordingly!

File: src/utils/narrow.js (hunk 20 above)
Comment (kunall17) [inline src/utils/narrow.js]: Used it before, not needed now!

File: src/message/renderMessages.js (hunk 15 above)
Comment (kunall17) [inline src/message/renderMessages.js]: Outbox uses id as the timestamp, hence if I use id and not timestamp then i'll have to put conditions for timestamp, if you want we can have both id and timestamp then no conditions?

**Comment by Sam1301** [comment]
Nice, looks good overall! 👍 
Not sure if this case could be handled in this pr as well, but when I turn the internet connection off and then back on; on sending consequent messages, the activity indicator in message tags is always visible.
Also, does this pr address the duplicate keys issue? I can repro it as well.

**Comment by kunall17** [comment]
@Sam1301 couldn't reproduce the bug, any particular steps?

**Comment by Sam1301** [comment]
Sure, I took the following steps for the indicator bug on an iOS simulator
1) turn internet connection off, then back on
2) send a message

At this point, I could see the loading indicators for the recently sent message always visible.

**Comment by kunall17** [comment]
I should have been more clear, I meant steps  for the `duplicate keys issue?`

**Comment by Sam1301** [comment]
Sure 😅 
the keys issues seems a bit hard to repro, but after doing the following steps a couple of times I got it again
1) sent some messages in narrow to self
2) switched to All PM messages narrow
2) if I didn't get the error in All PM narrow, I tried to reload and switched repeatedly between `narrow to self` and `All PM narrow`.

This was the warning message

```
Warning: flattenChildren(...): Encountered two children with the same key, `1:$269863`. Keys should be unique so that components maintain their identity across updates. Non-unique keys may cause children to be duplicated and/or omitted — the behavior is unsupported and could change in a future version.
```

**Comment by kunall17** [comment]
Thanks @Sam1301 i'll be trying these steps

**Comment by borisyankov** [comment]
I also did reproduce the warning when on 'Private Messages'.

**Comment by kunall17** [comment]
Ok, so now i understood what's going wrong 

So two events are being received one of [MESSAGE_FETCH_SUCCESS](https://user-images.githubusercontent.com/12700799/29442326-66062716-83ee-11e7-8d4f-4c960cc45e95.png) and one of [EVENT_NEW_MESSAGE](https://user-images.githubusercontent.com/12700799/29442324-65d3d9b4-83ee-11e7-9c42-3b37a293f771.png)

And both of them add a message with the same ID and hence duplicate key's are produced! 

Final state is [this](https://user-images.githubusercontent.com/12700799/29442325-65d50c62-83ee-11e7-95b7-4112e90bb650.png)
So likely it's not a bug with the outbox implementation! 

@borisyankov should I check for existing id's while adding in the reducer, or maybe you'll fix this?

**Comment by kunall17** [comment]
7584c26 prevents the same id messages adding to the state through EVENT_NEW_MESSAGE

Now the problem is the MESSAGE_FETCH_START is being triggered very frequently hence many a times the message which is sent to the server received through MESSAGE_FETCH_SUCCESS and then EVENT_NEW_MESSAGE is not received hence the indicator keeps showing, and the message remains in the outbox!

Not sure if this theory is 100% correct, but seems to be the problem!

File: src/message/headers/MessageHeader.js [hunk 21]
```diff
@@ -93,7 +93,7 @@ export default class MessageHeader extends PureComponent {
 
     if (message.type === 'private' && !isPrivateOrGroupNarrow(narrow) && !isTopicNarrow(narrow)) {
       const recipients =
-        message.display_recipient.length > 1
+        Array.isArray(message.display_recipient) && message.display_recipient.length > 1
```
Comment (borisyankov) [inline src/message/headers/MessageHeader.js]: Isn't `message.display_recipient` always an array if `type` is `private`?

File: src/chat/Chat.js [hunk 22]
```diff
@@ -32,7 +32,9 @@ export default class Chat extends PureComponent {
   };
 
   handleSend = () => {
-    if (this.listComponent.scrollToEnd) this.listComponent.scrollToEnd();
+    if (this.listComponent && this.listComponent.scrollToEnd !== undefined) {
```
Comment (borisyankov) [inline src/chat/Chat.js]: When is `scrollToEnd` undefined?

File: src/utils/narrow.js [hunk 23]
```diff
@@ -131,9 +132,17 @@ export const narrowFromMessage = (message: Message, email: string) => {
 
 export const extractTypeToAndSubjectFromNarrow = (
   narrow: Narrow,
+  users,
 ): { type: 'private' | 'stream', display_recipient: string, subject: string } => {
   if (isPrivateOrGroupNarrow(narrow)) {
-    return { type: 'private', display_recipient: narrow[0].operand, subject: '' };
+    return {
+      type: 'private',
+      display_recipient: narrow[0].operand.split(',').map(item => {
```
Comment (borisyankov) [inline src/utils/narrow.js]: That looks like a function to be extracted.

File: src/message/renderMessages.js (hunk 15 above)
Comment (borisyankov) [inline src/message/renderMessages.js]: Yes, we can have both.

File: src/chat/chatReducers.js [hunk 24]
```diff
@@ -95,19 +95,26 @@ export default (state: ChatState = initialState, action: Action) => {
...
     case EVENT_NEW_MESSAGE: {
-      return {
+      let stateChange = false;
+      const newState = {
         ...state,
         messages: Object.keys(state.messages).reduce((msg, key) => {
           const isInNarrow = isMessageInNarrow(action.message, JSON.parse(key), action.ownEmail);
-          msg[key] = isInNarrow ? [...state.messages[key], action.message] : state.messages[key];
+          const isMessageAlreadySaved = isInNarrow
+            ? state.messages[key].find(item => action.message.id === item.id) !== undefined
```
Comment (borisyankov) [inline src/chat/chatReducers.js]: That whole function can be improved, looks a bit unclear.
In fact the best way to refactor it, is to extract the merge code from the `MESSAGE_FETCH_SUCCESS` and call it with just one parameter.

**Comment by borisyankov** [comment]
The PR is moving along very well, but there is more work to be done.
Keep in mind it is high-priority! I'll also be tweaking the message list more to improve performance.

File: src/message/headers/MessageHeader.js (hunk 21 above)
Comment (kunall17) [inline src/message/headers/MessageHeader.js]: Nope, if its a private message not a grouped one its a string

File: src/chat/chatReducers.js (hunk 24 above)
Comment (kunall17) [inline src/chat/chatReducers.js]: I refactored the code 

EVENT_NEW_MESSAGE is very different from MESSAGE_FETCH_SUCCESS, hence did not take code from there!

**Comment by kunall17** [comment]
@borisyankov Done with the changes and refactored the EVENT_NEW_MESSAGE chatReducer a bit!

**Comment by borisyankov** [comment]
Cool. Another epic PR is ready to merge.
Keep in mind, more work on the outbox is likely needed (as the message list is not outbox aware anymore and might have broken something)
